### ISSUE TITLE
Identical linkages during generation

### ISSUE THREAD

**Issue by linas** [opencog/link-grammar#1348]
When asking the generator to create lots of linkages, some (many?) of them appear to be identical. I don't understand why.

FYI, I added flags to `link-generator`: it now creates 500 linkages by default, but prints only the top 20 of them.  The -k and -c flags, respectively.

**Comment by ampli** [comment]
> When asking the generator to create lots of linkages, some (many?) of them appear to be identical. I don't understand why.

I will try to exactly trace the way the apparently identical links are extracted.

> The -k and -c flags, respectively.

I noted that it might enter an infinite loop, depending on the values of -k and -c. Maybe it happens when there are not enough linkages, but I haven't investigated it further.

**Comment by linas** [comment]
I am also seeing many duplicate linkages with just the parser (i.e. NOT the generator)  Sometimes dozens or even hundreds that are all the same.  I get the feeling this has something to do with optional connectors (but this is a wild guess) So, for example, in my dict, I have structures that look like this:
```
(A+ or B+ or C+ or ()) and (A+ or B+ or C+ or ()) and (A+ or B+ or C+ or ())
```
but if my parse only needs one A link, I'll get three identical parses from the combinatorial possibilities. 

However, this is just a guess; when I saw this with the generator, I wasn't using optional connectors in this way; it might have been something else.

**Comment by ampli** [comment]
> Does that make sense?

The disjuncts for all these 3 cases should be the same.

>Maybe eliminate_duplicate_disjuncts() isn't working?

If you do `!!testit//` you will see that the duplicates are eliminated:
``` text
 linkparser>!!testit//
Token "testit" matches:
    testit                           64 disjuncts


Token "testit" disjuncts:
    testit                           39/63 disjuncts

          testit: [0] 0.000=  <> A+ A+
          testit: [1] 0.000=  <> A+ B+
          testit: [2] 0.000=  <> A+ C+
          testit: [3] 0.000=  <> A+
          testit: [4] 0.000=  <> A+ C+ A+
          testit: [5] 0.000=  <> A+ C+ B+
          testit: [6] 0.000=  <> A+ C+ C+
          testit: [7] 0.000=  <> A+ B+ A+
          testit: [8] 0.000=  <> A+ B+ B+
          testit: [9] 0.000=  <> A+ B+ C+
          testit: [10] 0.000=  <> A+ A+ A+
          testit: [11] 0.000=  <> A+ A+ B+
          testit: [12] 0.000=  <> A+ A+ C+
          testit: [13] 0.000=  <> B+ A+
          testit: [14] 0.000=  <> B+ B+
          testit: [15] 0.000=  <> B+ C+
          testit: [16] 0.000=  <> B+
          testit: [17] 0.000=  <> B+ C+ A+
          testit: [18] 0.000=  <> B+ C+ B+
          testit: [19] 0.000=  <> B+ C+ C+
          testit: [20] 0.000=  <> B+ B+ A+
          testit: [21] 0.000=  <> B+ B+ B+
          testit: [22] 0.000=  <> B+ B+ C+
          testit: [23] 0.000=  <> B+ A+ A+
          testit: [24] 0.000=  <> B+ A+ B+
          testit: [25] 0.000=  <> B+ A+ C+
          testit: [26] 0.000=  <> C+ A+
          testit: [27] 0.000=  <> C+ B+
          testit: [28] 0.000=  <> C+ C+
          testit: [29] 0.000=  <> C+
          testit: [30] 0.000=  <> C+ C+ A+
          testit: [31] 0.000=  <> C+ C+ B+
          testit: [32] 0.000=  <> C+ C+ C+
          testit: [33] 0.000=  <> C+ B+ A+
          testit: [34] 0.000=  <> C+ B+ B+
          testit: [35] 0.000=  <> C+ B+ C+
          testit: [36] 0.000=  <> C+ A+ A+
          testit: [37] 0.000=  <> C+ A+ B+
          testit: [38] 0.000=  <> C+ A+ C+
```

For more complex cases you can textually sort the disjunct text  (e.g `sort -u` should give you the same number of disjuncts, or `uniq -d` after sorting should give nothing).

(`-verbosity=102` also prints disjuncts, but only after `expression_prune()` and then again after `power_prune()`.)

**Comment by ampli** [comment]
Note that if you have linkage sampling, it may sometimes sample the same disjuncts again.

**Comment by linas** [comment]
> Note that if you have linkage sampling, it may sometimes sample the same disjuncts again.

Ahh! That might explain things ... the last one had 148K parses, and, with the default linkage limit of 1000, the results looked normal. But then I set `!limit=50000` and started seeing duplicates, sometimes 2 in a row, sometimes 14 in a row.  I will explore some more...

**Comment by linas** [comment]
Yep. For parsing, that's definitely the answer: random sampling of linkages. At least in the case of parsing.  When `!lim` is much smaller than the number of linkages, then the chance of getting duplicates is tiny. When it is greater than the number of linkages, there are no duplicates. When its 1/3rd, then many of the samples are duplicates.

This is a "drawing marbles from an urn with replacement" problem, it will tell you what the chances are of drawing the same marble repeatedly.

Perhaps the same idea explains generation, although I was trying to generate all possibilities.  I'll retry generation later...

**Comment by linas** [comment]
Closing -- Fixed in #1378 -- it was entirely due to random sampling. When When the linkage limit was just below the number of parses, almost half of the linkages were duplicates.
